keps/sig-auth/3299-kms-v2-improvements/README.md

```diff
@@ -0,0 +1,524 @@
++    KMSv2 *KMSv2Configuration `json:"kmsv2,omitempty"`
+ }
+ 
+ // AESConfiguration contains the API configuration for an AES transformer.
+@@ -98,3 +99,10 @@ type KMSConfiguration struct {
+	 // +optional
+	 Timeout *metav1.Duration `json:"timeout,omitempty"`
+ }
++
++type KMSv2Configuration struct {
++    // name is the name of the KMSv2 plugin to be used.
++    Name string `json:"name"`
++    // endpoint is the gRPC server listening address, for example "unix:///var/run/kmsv2-provider.sock".
++    Endpoint string `json:"endpoint"`
++}
+```
+
+Support key hierarchy in KMS plugin that generates local KEK and add v2alpha1 `KeyManagementService` proto service contract in Kubernetes to include `currentKeyID`, `metadata`, and `status`. 
+
+Key Hierarchy in KMS plugin (reference implementation):
+
+1. No changes to the API server, keep 1:1 DEK mapping
+    1. Assumption: A KMS plugin that was implemented using a local HSM would not need any changes because it would be able to handle the amount of encryption calls with ease since it would not need to perform network IO
+    1. Assumption: local gRPC calls to the KMS plugin do not represent significant overhead
+1. KMS plugin generates its own local KEK in-memory
+1. External KMS is used to encrypt the local KEK
+1. Local KEK is used for encryption of DEKs sent by API server
+1. Local KEK is used for encryption based on policy (N events, X time, etc)
+
+Since key hierarchy is implemented at the KMS plugin level, it should be seamless for the kube-apiserver. So whether the plugin is using a key hierarchy or not, the kube-apiserver should behave the same.
+
+What is required of the kube-apiserver is to be able to tell the KMS plugin which KEK (local KEK or KMS KEK) it should use to decrypt the incoming DEK. To do so, upon encryption, the KMS plugin could provide the encrypted local KEK as part of the `metadata` field in the `EncryptResponse`. The kube-apiserver would then store it in etcd next to the DEK. Upon decryption, the kube-apiserver provides the encrypted local KEK in `metadata` and `observedKeyID` from the last encryption when calling Decrypt. In case no encrypted local KEK is provided in the `metadata`, then we can assume key hierarchy is not used. The KMS plugin would query the external KMS to use the remote KEK to decrypt the DEK (same behavior as today). No state coordination is required between different instances of the KMS plugin.
+
+For the reference KMS plugin, the encrypted local KEK is stored in etcd via the `metadata` field, and once decrypted, it can be stored in memory as part of the KMS plugin cache to be used for encryption and decryption of DEKs. The encrypted local KEK is used as the key and the decrypted local KEK is stored as the value.
+
+```proto
+message EncryptResponse {
+    // The encrypted data.
+    bytes ciphertext = 1;
+    // The KMS key ID used for encryption operations.
+    // This can be used to drive rotation.
+    string currentKeyID = 2;
```

Comment:
It's strange but I think we use snake case for non-generated proto APIs.

Reply:
Interesting, that does at least line up with CRI https://github.com/kubernetes/cri-api/blob/1c90a6f7abe8f5b7331c95960a5106c9876a9e4f/pkg/apis/runtime/v1/api.proto#L272